operator/src/command_utils.rs

```diff
@@ -42,138 +57,279 @@ impl CommandType {
         }
     }
 
-    /// Return the current command status
-    pub fn get_status(&self) -> Option<CommandStatus> {
-        match self {
-            CommandType::Restart(restart) => restart.status.clone(),
-            CommandType::Start(start) => start.status.clone(),
-            CommandType::Stop(stop) => stop.status.clone(),
-        }
-    }
-
-    /// Writes / updates the status in the command object
-    ///
-    /// # Arguments
-    /// * `client` - Kubernetes client
-    /// * `new_status` - Desired new status
-    ///
-    pub async fn write_status(
-        &self,
-        client: &Client,
-        new_status: &CommandStatus,
-    ) -> OperatorResult<bool> {
-        match self {
-            CommandType::Restart(restart) => {
-                update_status(client, restart, &self.get_status(), new_status).await
-            }
-            CommandType::Start(start) => {
-                update_status(client, start, &self.get_status(), new_status).await
-            }
-            CommandType::Stop(stop) => {
-                update_status(client, stop, &self.get_status(), new_status).await
-            }
-        }
-    }
-
-    /// Implementation of command behavior when starting the command
+    /// Implementation of command behavior when starting to execute the command
     /// e.g. delete pods for restart
     ///
     /// # Arguments
     /// * `client` - Kubernetes client
+    /// * `cluster` - Spark cluster custom resource
+    /// * `current_command` - Command that is currently processed
     /// * `pods` - All available cluster pods
     ///
     pub async fn process_command_execute(
         &self,
         client: &Client,
+        cluster: &SparkCluster,
+        current_command: &CurrentCommand,
         pods: &[Pod],
-    ) -> OperatorResult<bool> {
-        let new_status = &CommandStatus {
-            started_at: get_time(),
-            finished_at: None,
-            message: Some(CommandStatusMessage::Started),
-        };
-
-        match self {
+    ) -> OperatorResult<ReconcileFunctionAction> {
+        info!(
+            "Executing [{}] command '{}'",
+            current_command.command_type, current_command.command_ref,
+        );
+        // set the current_command in the cluster custom resource status
+        let updated_cluster = update_current_command(client, cluster, current_command).await?;
+        // apply command specific action
+        return match self {
             CommandType::Restart(_) => {
                 for pod in pods {
                     client.delete(pod).await?;
                 }
-
-                self.write_status(client, new_status).await
+                update_cluster_status(client, &updated_cluster, &ClusterStatus::Running).await?;
```

Comment:
I know it was my suggestion but should we have more statuses?

Running, Starting, Stopping, Restarting, Stopped?
In theory we should probably also expose those as conditions, could you add a TODO to update_cluster_status?

Reply:
I thought about that too. But they are different things in my opinion.
Stopped for me is basically a flag which signals to not act on anything.
Stopping is the process of the "Stop" command running. 

I would like to avoid mixing that and e.g. put Stopping in the conditions?

Reply:
Okay, but in that case `clusterStatus` is not a good term.
It's more....I don't know.... desired status or something.
Maybe ask the others for opinions?

Reply:
I think desired status would not fit (at least how its currently implemented). Desired is something you work towards, e.g. if a restart command arrives, the desired status would still be running?
What about something like clusterPhase similiar to pods?

Reply:
Paging @soenkeliebau 
Any opinions?

This comes back to another thing I stumbled across while listening to a Podcast with Kelsey Hightower: I'm wondering if "stop" should be part of the spec. "status: stopped" seems like a specworthy thing because it's a change in desired state while restart just transitions from started to started.

I don't really like `clusterPhase` but can't say exactly why that is.

Reply:
I have no answers, just some musings and opinions that I am of course happy to share :)

Regarding stopped and stopping - I'm not sure we need an extra state stopping, isn't that described by "stopped but there are still pods alive"?
Whereas "stopped an no pods" is the stable state that we want to achieve..

On stop being part of the spec - I can see the argument why that would be correct from a theoretical point of view, but I hesitate if we don't fall into the same fallacy that Kubernetes did here and there of doing stuff right from an academic/theoretical point of view.
In day-to-day life this would mean that stopping a cluster changes the state which you probably have in git somewhere and which most probably is under some restriction to who can change this when and would trigger a sync with the CRDs in Kubernetes. That is not really feasible I think. But if the restart operator makes this change then the state in version control now is different from the actual state - we'd need a backwards sync mechanism.
I'd be happy to include the started/stopped information in an ephemeral state that would not be replicated if you spin up the same cluster on a different Stackable platform. To me this could be treated like for exapmle the fact whether or not the hdfs has been formatted. Yes, that is state, but would we include it in the spec or would that be state?
If we take the spec and push that to a new system, the fs would need to be formatted again - just like the cluster would need to be started again. 

Regarding the name, which I think was the actual question: clusterStatus would work for me I think..but it is a fairly broad term that could potentially mean lots of other things as well.. clusterRunState / cluster ExecutionState maybe ?

Reply:
I think Sönke is right, i would not put it in the spec either. For the name, i think clusterExecutionState is on point.

Reply:
Maybe a few words on my "ephemeral state" - the idea is loosely based on Elasticsearch, where there is two different ways to change settings: transient and persistent. A transient setting will vanish after a full cluster restart (all nodes stopped at the same time) whereas a persistent setting will be stored on disk and survive even a full restart.

I mentally mapped the transient state to anything influencing a specific cluster that was deployed on a specific instance of stackable. For this one cluster we should not loose the information if it it stopped or not.

A full restart would be taking the spec that deployed this cluster and sending it to a different Stackable platform with different backing agents to deploy a totally new cluster but with the same settings.
If we loose the fact that the cluster was stopped for this action I'd be fine with that.

Even more, for the "HDFS formatted" state we actually have to loose that, because on the new cluster it has not been formatted, we need to do that.

Writing this down I can see that it is a very convoluted comparison, but maybe I made my thoughts a bit clearer?


fwiw I do agree with you that "stopped" should ideally be part of the spec, but I don't like designing something that has known issues for which we have no answer, just to be "correct". We have always said that we would compromise in the interest of usability - I think this might be one of those times ..

Reply:
Okay. I understand.
HDFS formatted - for me - is not state but a thing the operator needs to do to achieve the goal of a running HDFS cluster.

HDFS = stopped, for me is state and I think I'd like to pass that on to other clusters.

In short: I think I disagree with the conclusion - but understand where you're coming from - and we can keep it like it is for now.

I still very much disagree about "stopped" vs "stopping" though.